File: Data Structure/5. Heap/heap.py
# a python implementation of min heap

# min heap class
class MinHeap:
  # constructor
  def __init__(self, size):
    self.capacity = size
    self.heap_array = [None for i in range(size)]
    self.heap_size = 0
  
  # check if empty
  def is_empty(self):
    return self.heap_size == 0

  # output the empty status
  def heap_status(self):
    if self.is_empty():
      print("The heap is empty")
    else:
      print("The heap is not empty")
    print # new line for better looking output

  # given an index, get parent index
  def parent(self, index):
    return (index - 1) / 2

  # given an index, get left child index
  def left(self, index):
    return (2 * index) + 1

  # given an index, get right child index
  def right(self, index):
    return (2 * index) + 2

  # return the minimum node from min heap
  def get_min(self):
    return self.heap_array[0]

  # get size of the heap
  def get_size(self):
    return self.heap_size

  # display the content of min heap
  def dispaly(self):
    if self.is_empty:
      self.heap_status

    for i in range(self.heap_size):
      print(str(self.heap_array[i])),
    print

  # insert new value to the min heap
  def insert(self, value):
    # the heap is full
    if self.heap_size == self.capacity:
      print("\nOverflow: Heap is full already!")
      return
    
    # insert the new key at the end
    self.heap_size += 1
    index = self.heap_size - 1
    self.up_heap(index, value)
    
  # decrease value of key at given index to new value
  def up_heap(self, index, value):
    # retain the min heap order by swapping parent with current key
    heap_array = self.heap_array
    heap_array[index] = value
    while index != 0 and heap_array[self.parent(index)] > heap_array[index]:
      # swap parent and the current node if parent > current value
      temp = heap_array[index]
      heap_array[index] = heap_array[self.parent(index)]
      heap_array[self.parent(index)] = temp
      index = self.parent(index) # move up the index to check min heap property

  # to heapify a subtree with the root at given index
  def min_heapify(self, index):
    # get left and right child index
    leftIdx = self.left(index)
    rightIdx = self.right(index)

    # find minIdx by comparing left or right with current node value
    minIdx = index
    heap_array = self.heap_array
    if leftIdx < self.heap_size and heap_array[leftIdx] < heap_array[index]:
      minIdx = leftIdx
    if rightIdx < self.heap_size and heap_array[rightIdx] < heap_array[minIdx]:
      minIdx = rightIdx

    # if the minIdx has been modified, swap the current value with value at minIdx
    if minIdx != index:
      heap_array[index], heap_array[minIdx] = heap_array[minIdx], heap_array[index]
      
      # recursively call heapify with new minIdx to create min heap
      self.min_heapify(minIdx)

  # remove minimum node from min heap
  def remove_min(self):
    # edge case: empty or only one element exists
    if self.get_size <= 0:
      print("Empty! No element to extract")
      return
    if self.get_size == 1:
      self.heap_size -= 1

    # store the min value, move lower right node to the top
    self.heap_array[0] = self.heap_array[self.heap_size-1]
    self.heap_size -= 1

    # call min_heapify with root node to restore the heap structure
    self.min_heapify(0)

# ============================
# testing the min heap class
# ============================

# create a empty heap with size 5
minHeap = MinHeap(5)
print("\n- Min Heap Created")
minHeap.heap_status()

# insert new values to the heap
minHeap.insert(3)
minHeap.insert(2)
minHeap.insert(5)
minHeap.insert(4)
minHeap.insert(1)
print("- Inserted values: 3 2 5 4 1")
minHeap.heap_status()

# display the heap content
print("- Display the min heap content:"),
minHeap.dispaly()
print

# get minimum value of the min heap
print("- Get the min value: " + str(minHeap.get_min()))
print

# get size  of the min heap
print("- Get the size: " + str(minHeap.get_size()))
print

# remove min value of the min heap object
minHeap.remove_min()
print("- After removing minimum value of min heap: "),
minHeap.dispaly()
print

File: Data Structure/5. Heap/heap.cpp
// A C++ program to demonstrate common Binary Heap Operations
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

// A class for Min Heap
class MinHeap {
  vector<int> heapArray; // array representation of heap
  int capacity; // maximum possible size of min heap
  int heap_size; // Current number of elements in min heap

  public:
  // Constructor to initialize heap properties
  MinHeap(int siz) {
    heap_size = 0;
    capacity = siz;
    heapArray.resize(siz);
  }

  // is empty?
  bool isEmpty() { return heap_size == 0; }

  // get parent index of given index i
  int parent(int i) { return (i-1) / 2; }

  // get index of left child of node at index i
  int left(int i) { return (2 * i + 1); }

  // get index of right child of node at index i
  int right(int i) { return (2 * i + 2); }

  // return the minimum key (key at root) from min heap
  int getMin() { return heapArray[0]; }

  // get size of the heap
  int getSize() { return heap_size; }

  // inserts a new value
  void insertKey(int value) {
    if (heap_size == capacity){
      cout << "\nOverflow: Could not insert anymore\n";
      return;
    }

    // insert the new key at the end
    heap_size++;
    int i = heap_size - 1;
    heapArray[i] = value;

    // swap to fix the min heap property if it is violated
    while (i != 0 && heapArray[parent(i)] > heapArray[i]) {
      // swap the parent and the current key is parent > current key
      int temp = heapArray[i];
      heapArray[i] = heapArray[parent(i)];
      heapArray[parent(i)] = temp;
      i = parent(i); // move up the index to check min heap property
    }
  }

  // change value of key at index 'i' to new_val
  void decreaseKey(int i, int new_val) {
    // retain the min heap order by swapping parent with current key
    heapArray[i] = new_val;
    while (i != 0 && heapArray[parent(i)] > heapArray[i]) {
      // swap the parent and the current key is parent > current key
      int temp = heapArray[i];
      heapArray[i] = heapArray[parent(i)];
      heapArray[parent(i)] = temp;
      i = parent(i); // move up the index to check min heap property
    }
  }

  // to heapify a subtree with the root at given index
  void minHeapify(int i) {
    // get left and right child index
    int minIdx = i;
    int leftIdx = left(i);
    int rightIdx = right(i);

    // find minIdx by comparing left and child index with current index value
    if (leftIdx < heap_size && heapArray[leftIdx] < heapArray[i])
        minIdx = leftIdx;
    if (rightIdx < heap_size && heapArray[rightIdx] < heapArray[minIdx])
        minIdx = rightIdx;

    // if the minIdx has been changed, swap the current value with minIdx value
    if (minIdx != i) {
      swap(heapArray[i], heapArray[minIdx]);

      // recursively call heapify with new minIdx to create min heap 
      minHeapify(minIdx);
    }
  }

  // remove minimum element (or root) from min heap
  void removeMin() {
    if (getSize() <= 0)
      cout << "Empty! No element to extract!";
    if (heap_size == 1)
      heap_size--;

    // store the minimum value, move lower right element to top
    heapArray[0] = heapArray[heap_size-1];
    heap_size--;
    
    // call minHeapif() with root node to restore the heap structure
    minHeapify(0);
  }

  // deletes a key stored at index i
  void deleteKey(int i) {
    decreaseKey(i, INT_MIN); // move up the element to delete
    removeMin(); // remove the top
  }

  // display heap array
  void display() {
    for (int i = 0; i < heap_size; i++)
      cout << heapArray[i] << " ";
    cout << endl;
  }
};

// testing the min heap class
int main() {
  cout << "Testing min heap implementation: \n" << endl;

  // create min heap object
  MinHeap heap(5);
  cout << "- Created min heap object with size 5\n" << endl;

  // insert values into the heap
  heap.insertKey(3);
  heap.insertKey(2);
  heap.insertKey(5);
  heap.insertKey(4);
  heap.insertKey(1);
  cout << "- Inserted values: 3 2 5 4 1\n" << endl;

  // display the inside of heap array
  cout << "- Display the content of min heap: ";
  heap.display();
  cout << endl;

  // get minimum element of the min heap
  cout << "- Get minimum value of min heap: ";
  cout << heap.getMin() << "\n" << endl;
  
  // remove minimum value of min heap
  heap.removeMin();
  cout << "- After removing minimum value of min heap: ";
  heap.display();
  cout << endl;

  // delete a given key
  heap.deleteKey(5);
  cout << "- After deleteting key 5: ";
  heap.display();
  cout << endl;

  // insert value 1
  heap.insertKey(1);
  cout << "- Inserted value 1: ";
  heap.display();
  cout << endl;
}